add nft_metadata_update event to NEP171 (#469)
near/NEPs: neps/nep-0171.md

> 

Reviewer: Shouldn't we have a changelog?

Reviewer 2: Do you refer to the NEP changelog? There is a placeholder there already. We will fill it out once we have all the reviews, so we can compile all the feedback and the final decision details into it.

Reviewer: Yes. wouldn't it be better to have it already set and be part of the final review?

Reviewer 2: @robert-zaremba Ideally, yes, but I cannot see how to make it productive using GitHub UI. Feel free to review [my SME review comment](https://github.com/near/NEPs/pull/469#issuecomment-1476817621).

### Events

> NEAR and third-party applications need to track `mint`, `update`, `transfer`, `burn`, and `contract_metadata_update` events for all NFT-driven apps consistently.

Reviewer 3: A token is updated if the `owner_id` is changed, because `owner_id` is one of the Token fields: https://nomicon.io/Standards/Tokens/NonFungibleToken/Metadata. 
So, for the transfer, we have to emit 2 logs, which is weird.

Do we want to use `token_metadata_update`?

Author: This is a very good point. What about `nft_metadata_update` to keep the naming convention for NFT-related events?

Author: Just updated to `nft_metadata_update`

> NEAR and third-party applications need to track `mint`, `update`, `transfer`, `burn`, and `contract_metadata_update` events for all NFT-driven apps consistently.

Reviewer: Yes, let's do normal formatting. We describe an action here, rather than an object.

This extension addresses that.

Keep in mind that applications, including NEAR Wallet, could require implementing additional methods to display the NFTs correctly, such as [`nft_metadata`][Metadata] and [`nft_tokens_for_owner`][NFT Enumeration]).

#### Events Interface

Non-Fungible Token Events MUST have `standard` set to `"nep171"`, standard version set to `"1.2.0"`, `event` value is one of `nft_mint`, `nft_burn`, `nft_transfer`, `nft_metadata_update`, or `contract_metadata_update`, and `data` must be of one of the following relevant types: `NftMintLog[] | NftTransferLog[] | NftBurnLog[] | NftMetadataUpdateLog[] | NftContractMetadataUpdateLog[]`:

```ts
interface NftEventLogData {
    standard: "nep171",
    version: "1.2.0",

>     event: "nft_mint" | "nft_burn" | "nft_transfer" | "nft_metadata_update" | "contract_metadata_update",

Reviewer: Side comment for the future events (not the NEP-171, as it is already set).
Personally, I don't think we need to have `nft_` prefix in all the event names (not the struct names... they are not relevant to the standard). The NFT namespace is already implicity with the `standard: "nep171"`.

Author: I disagree, I think this prefix makes it easier for indexers to handle events by type, and it helps categorize events within the contract.

    memo?: string
}

// An event log to capture token updating (e.g. new metadata)
// Arguments
// * `token_ids`: ["1", "abc"]
// * `memo`: optional message
interface NftUpdateLog {

>     token_ids: string[],

Reviewer 3: Have we considered the option to include the updated NFT in the logs?
At the first glance, I see obvious pros - no need to make the request to see updated NFT, and obvious cons - the log is growing, while we are trying to save the letters. 
I don't have the answer here, just want to know your opinion.

Reviewer 2: @telezhnaya It was discussed on the gov forum (see the link in the PR description). The major concern I see with posting more details into the logs is that the media might not fit into the logs. I think we are inevitably heading into the direction where we will use the logs to quickly catch if a relevant change happened, but then to learn more details you would need to execute some view function (as an optimization, we could even consider following the state updates on the indexer side, and execute view functions locally)

Reviewer 3: Ah, I see. I totally forgot about the media, you are right.

    memo?: string
}

> // An event log to capture token metadata updating

Reviewer: ```suggestion
// An event emitted when token metadata is updated. 
```

Reviewer: I don't think _log_ is appropriate here. This is an object, not a log.

Author: I don't think it's a big deal, probably easier to just keep it consistent with existing comments. I personally don't find it confusing.

Reviewer: logging and event != emitting an event. In blockchian context we talk about emitting an event usually. So if we want to be consistent with that, then we should use _emit_

Reviewer 2: @robert-zaremba Let's create a separate PR with this improvement to the terminology, so we can update all the Events NEPs at once and discuss it separately from this NEP.

// Arguments
// * `token_ids`: ["1", "abc"]
// * `memo`: optional message
interface NftMetadataUpdateLog {
    token_ids: string[],

>     memo?: string

Reviewer: in my opinion this should not be needed - this is already in the transaction, and I don't think memo is something that should be indexed. However since this is in other NEP-171 events, we can keep it here.

Reviewer 2: `memo` is not a necessary part of the transaction. There is even no function associated with this event as there is no meaningful one.
